cmd/release-controller/http.go

```diff
@@ -925,3 +1015,248 @@ var extendedRelTime = []humanize.RelTimeMagnitude{
+							Name:nextReleaseName,
+						},
+						time.Unix(ts,0).Format(time.RFC3339),
+						tag,
+					})
+				}
+			}
+		}
+		sort.Slice(candidates, func(i, j int) bool {
+			return candidates[i].CreationTime > candidates[j].CreationTime
+		})
+		releaseCandidates[stream] = &ReleaseCandidateList{Items:candidates}
+	}
+	return releaseCandidates, nil
+}
+
+func (c *Controller) filterByUpgradeSuccess(rcList map[string]*ReleaseCandidateList, successPercent float64) map[string]*ReleaseCandidateList {
+	for stream, list := range rcList {
+		for i, candidate := range list.Items {
+				upgradeSuccess := make([]string, 0)
+				upgrades := c.graph.UpgradesTo(candidate.FromTag)
+				for _, u := range upgrades {
+					if u.Total > 0 {
+						if float64(100*u.Success)/float64(u.Total) > successPercent {
+							upgradeSuccess = append(upgradeSuccess, u.From)
+						}
+					}
+				}
+				sort.Strings(upgradeSuccess)
+				rcList[stream].Items[i].UpgradeFrom = upgradeSuccess
+		}
+	}
+	return rcList
+}
+
+func (c *Controller) httpReleaseCandidateList(w http.ResponseWriter, req *http.Request) {
+	start := time.Now()
+	defer func() { glog.V(4).Infof("rendered in %s", time.Now().Sub(start)) }()
+	vars := mux.Vars(req)
+	releaseStreamName := vars["release"]
+	releaseCandidateList, err := c.getReleaseCandidateList(strings.Split(releaseStreamName, ",")...)
+	if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+	}
+	releaseCandidateList = c.filterByUpgradeSuccess(releaseCandidateList, 0.0)
+
+	switch req.URL.Query().Get("format") {
+	case "json":
+		data, err := json.MarshalIndent(&releaseCandidateList, "", "  ")
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		fmt.Fprintf(w, string(data))
+	default:
+		fmt.Fprintf(w, htmlPageStart, "Release Status")
+		page := template.Must(template.New("candidatePage").Funcs(
+			template.FuncMap{
+				"nextReleaseName": func (list *ReleaseCandidateList) string {
+					if list == nil || list.Items == nil || len(list.Items) == 0 {
+						return "next release"
+					}
+					return list.Items[0].Name
+				},
+			},
+		).Parse(candidatePageHtml))
+
+		if err := page.Execute(w, releaseCandidateList); err != nil {
+			glog.Errorf("Unable to render page: %v", err)
+		}
+		fmt.Fprintln(w, htmlPageEnd)
+	}
+}
+
+func (c *Controller) httpReleaseCandidate(w http.ResponseWriter, req *http.Request) {
+	start := time.Now()
+	defer func() { glog.V(4).Infof("rendered in %s", time.Now().Sub(start)) }()
+	vars := mux.Vars(req)
+	releaseStreamName := vars["release"]
+	releaseCandidateList, err := c.getReleaseCandidateList(strings.Split(releaseStreamName, ",")...)
+	if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+	}
+	releaseCandidateList = c.filterByUpgradeSuccess(releaseCandidateList, 0.0)
+
+	switch req.URL.Query().Get("format") {
+	default:
+		topCandidates := make(map[string]*ReleasePromoteJobParameters)
```

Comment:
Since we're going to use the annotation, I'm not sure we should merge this method as is.  I would maybe suggest simply moving this to another PR (verify keep) after you get the list functionality reviewed and merged (refactoring anything to share it).